pkg/webhook/creator/mutate.go

```diff
@@ -0,0 +1,140 @@
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+package creator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/che-incubator/che-workspace-operator/pkg/controller/workspace/model"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// CreatorChecker checks that every workspace-related deployment, pod (who has workspace id label) has creator
+// annotation and  it's not modified
+type CreatorChecker struct {
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+// CreatorChecker verify if upcoming object has creator annotation
+func (a *CreatorChecker) Handle(_ context.Context, req admission.Request) admission.Response {
+	if req.Kind.Kind == "Deployment" || req.Kind.Kind == "Pod" {
+		if req.Operation == v1beta1.Create {
+			return a.handleUnstructuredCreate(req)
+		} else if req.Operation == v1beta1.Update {
+			return a.handleUnstructuredUpdate(req)
+		}
+	}
+	return admission.Denied(fmt.Sprintf("This admission is not supposed to handle %s operation for %s. Let administrator to know about this issue", req.Operation, req.Kind))
+}
+
+func (a *CreatorChecker) handleUnstructuredCreate(req admission.Request) admission.Response {
+	o := &unstructured.Unstructured{}
+
+	err := a.decoder.Decode(req, o)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if o.GetAnnotations() == nil {
+		return admission.Denied(fmt.Sprintf("Workspace related %s must have creator annotation specified", req.Kind))
+	} else {
+		if _, ok := o.GetAnnotations()[model.WorkspaceCreatorAnnotation]; !ok {
+			return admission.Denied(fmt.Sprintf("Workspace related %s must have creator annotation specified", req.Kind))
+		}
+	}
+
+	return admission.Allowed("creator annotation is present")
+}
+
+func (a *CreatorChecker) handleUnstructuredUpdate(req admission.Request) admission.Response {
+	newObject := &unstructured.Unstructured{}
+	oldObject := &unstructured.Unstructured{}
+
+	err := a.decoder.Decode(req, newObject)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	err = a.decoder.DecodeRaw(req.OldObject, oldObject)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	updated, errMsg := a.checkAnnotations(oldObject.GetAnnotations(), newObject.GetAnnotations(), req.Kind.Kind)
+
+	if errMsg != "" {
+		return admission.Denied(errMsg)
+	}
+
+	if updated != nil {
+		return returnAnnotatedUnstructured(req.Object.Raw, newObject, updated)
+	}
+
+	return admission.Allowed("new deployment has the same creator as old one")
+}
+
+func (a *CreatorChecker) checkAnnotations(old, new map[string]string, kind string) (map[string]string, string) {
```

Comment:
Consider returning `(map[string]string, error)` instead, using `fmt.Errorf()`

Reply:
Actually I'm using `fmt.Sprintf` but not `fmt.Errorf()`. Putting error here - mean that we need to create an error object which then won't be used since `admission.Denied` requires string...
Leaving as-is and we'll change it later if needed.